arm64: dts: qcom: msm8916-samsung-gprime: Add gprimeltecan and fortuna3g

wonderfulShrineMaidenOfParadise opened the pull request:
Closes #225 

Samsung Galaxy Grand Prime is a smartphone using the MSM8916
SoC released in 2014.

The Grand Prime variants are very similar, with some differences in
accelerometer, NFC and panel. The common parts are shared in
msm8916-samsung-gprime-common.dtsi to reduce duplication.

Changes:
- GPIO keys and WCNSS
- Vibrator
- rt5033 fuelgauge. Also compatible with sec-fuelgauge
- Panel selection in lk2nd
- Zinitix BT541 with touchkey
- Modem and sound
- Audio jack

gprimeltecan:
- reserved memory
- BMC150 accelerometer
- S3FWRN5 NFC chip

fortuna3g:
- K2HH accelerometer
- s6d78a0-gh9607501a panel

- [ ] msm8916-mainline/lk2nd/pull/151
- [x] msm8916-mainline/linux-panel-drivers/pull/12

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add sound and modem

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add audio jack detection

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add accelerometer

This device has ST LIS2HH12 accelerometer.
Add support for it.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add vibrator

Motor is powered by VPH_PWR (aka Vbat) and controlled by GPIO.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add touch screen

Like msm8916-samsung-a3u-eur, the Grand Prime uses a Zinitix BT541
touch screen. Add it together with the necessary fixed-regulator to the
device tree.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add rt5033 battery

Like the Samsung Galaxy A3/A5, the Grand Prime uses a Richtek RT5033 PMIC
as battery fuel gauge, charger, flash LED and for some regulators.
For now, only add the fuel gauge/battery device to the device tree, so we
can check the remaining battery percentage.

The other RT5033 drivers need some more work first before they can be used
properly.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add display panel

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
HACK: thermal/drivers/tsens: gprime

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add NFC

Grand Prime SM-G530Y (fortunaltezt) has a NXP PN547, which is supported
by the nxp-nci-i2c driver in mainline. It seems to detect NFC tags
using "nfctool" just fine, although more testing is difficult given
there seem to be very few useful applications making use of the
Linux NFC subsystem. :(

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add display panel

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

buddyjojo committed:
arm64: dts: qcom: msm8916-samsung-gprimeltecan: Add accelerometer/magnetometer

Grand Prime Canada uses a Bosch BMC150 accelerometer/magnetometer combo.
The chip provides two separate I2C devices for the accelerometer
and magnetometer that are already supported by the bmc150-accel
and bmc150-magn driver

Signed-off-by: Joe Mason <buddyjojo06@outlook.com>

Commit:
squash! MSM8916: drm/panel: Generate using linux-mdss-dsi-panel-driver-generator

X-Code-Generator: msm8916-mainline/linux-panel-drivers@eff6d45 (dirty)
Signed-off-by: lmdpdg <lmdpdg@localhost>
Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

SiddharthManthan committed:
arm64: dts: qcom: msm8916-samsung-fortuna3g: Add initial device tree

Signed-off-by: Siddharth Manthan <siddharth.manthan@gmail.com>

SiddharthManthan committed:
arm64: dts: qcom: msm8916-samsung-fortuna3g: Add accelerometer

This device has ST LIS2HH12 accelerometer.
Add support for it.

Signed-off-by: Siddharth Manthan <siddharth.manthan@gmail.com>

SiddharthManthan committed:
arm64: dts: qcom: msm8916-samsung-fortuna3g: Add display panel

Signed-off-by: Siddharth Manthan <siddharth.manthan@gmail.com>

SiddharthManthan committed:
arm64: dts: qcom: msm8916-samsung-fortuna3g: Add sound and modem

Signed-off-by: Siddharth Manthan <siddharth.manthan@gmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprime-common: Add common device tree

The Samsung Galaxy Grand Prime is a phone based on MSM8916. It is similar
to the other Samsung devices based on MSM8916 with only a few minor
differences.

The device tree contains initial support for the Grand Prime with:
 - UART
 - eMMC/SD card
 - Buttons
 - WiFi/Bluetooth (WCNSS)
 - USB

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-gprimeltecan: Add initial device tree (v5)

v5: Rename from msm8916-samsung-gprime

Samsung Galaxy Grand Prime (SM-G530W) is a smartphone using the MSM8916
SoC released in 2014.

The Grand Prime variants are very similar, with some differences in
accelerometer and NFC. The common parts are shared in
msm8916-samsung-gprime-common.dtsi to reduce duplication.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add initial device tree

Samsung Galaxy Grand Prime (SM-G530Y) is a smartphone using the MSM8916
SoC released in 2014.

The Grand Prime variants are very similar, with some differences in
accelerometer and NFC. The common parts are shared in
msm8916-samsung-gprime-common.dtsi to reduce duplication.

Unfortunately, SM-G530Y was released with outdated 32-bit only
firmware and never received any update from Samsung. Since the 32-bit
TrustZone firmware is signed there seems to be no way currently to
actually boot this device tree on arm64 Linux at the moment. :(

However, it is possible to use this device tree by compiling an ARM32 kernel
instead. The device tree can be easily built on ARM32 with an #include
and it works really well there. To avoid confusion for others it is still
better to add this device tree on arm64. Otherwise it's easy to forget
to update this one when making some changes that affect all MSM8916 devices.

Maybe someone finds a way to boot ARM64 Linux on this device at some point.
In this case I expect that this device tree can be simply used as-is.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm: dts: qcom-msm8916-samsung-fortunaltezt: Include dts from arm64

After adding all necessary support for MSM8916 SMP/cpuidle without PSCI
on ARM32, build the Samsung Galaxy Grand Prime device tree from the arm64
tree together with the ARM32 include to allow booting this device on ARM32.

The approach to include device tree files from other architectures is
inspired from e.g. the Raspberry Pi (bcm2711-rpi-4-b.dts) where this is
used to build the device tree for both ARM32 and ARM64.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi:
```diff
@@ -78,6 +91,35 @@
+	panel: panel@0 {
+		compatible = "samsung,gprime-panel";
+		reg = <0>;
+		vreg-supply = <&pm8916_l6>;
```
`vreg-supply` sounds weird. I'd either call this supply `vddio-supply` or simply drop it (toss a coin if you want). [L6 is already enabled/disabled as needed for dsi0](https://github.com/msm8916-mainline/linux/blob/edd4ec00a14e682b8a2ad7090131138372f54d11/arch/arm64/boot/dts/qcom/msm8916-pm8916.dtsi#L12).

cc @SiddharthManthan for https://github.com/msm8916-mainline/linux-panel-drivers/pull/12 (please coordinate who will make this change)

wonderfulShrineMaidenOfParadise replied:
I would leave s6d78a0-gh9607501a in l-p-d to @SiddharthManthan. On the other hand I am working on hx8389c/hx8389c-gh9607501a.

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-gprime-common.dtsi:
```diff
@@ -87,6 +100,29 @@
+		touchscreen-size-y = <960>;
+
+		vdd-supply = <&reg_vdd_tsp_a>;
+		vddo-supply = <&pm8916_l6>;
```
Please use the new supply names (see https://github.com/msm8916-mainline/linux/commit/c54be0e32e54abdf7b89d56fe9edebc2f319acee). It should be:

```
vcca-supply = <&reg_vdd_tsp_a>;
vdd-supply = <&pm8916_l6>;
```

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-gprimeltecan.dts:
```diff
@@ -0,0 +1,29 @@
+			reg = <0x0 0x8c000000 0x0 0x200000>;
+			no-map;
+		};
+	};
```
The reserved-memory for gprimeltecan and fortuna3g seems identical, can we put it into the common dts?

NOTE: If fortunaltezt is different here this is exactly the situation where I need the full changes to be able to review properly. If yes, please push the fortunaltezt changes to this branch. I can skip them when merging if they are not fully ready but I would like to consider them in my review.

wonderfulShrineMaidenOfParadise replied:
Pushed fortunaltezt changes.

stephan-gh replied:
I'd put the reserved-memory in the common dtsi. You have two options (toss a coin again if you want):

- Use the 0x5a00000 for all gprimes (this will waste 2 MiB RAM on most variants)
- Use 0x5800000 in the common dtsi and override it for fortunaltezt like this:

```
	reserved-memory {
		/delete-node/ gps@8c000000;

		mpss_mem: mpss@86800000 {
			reg = <0x0 0x86800000 0x0 0x5a00000>;
			no-map;
		};

		gps_mem: gps@8c200000 {
			reg = <0x0 0x8c200000 0x0 0x200000>;
			no-map;
		};
	};
```

SiddharthManthan replied:
Overriding is preferred as ram is a luxury on fortuna3g.

SiddharthManthan replied:
Resolved. No need to commonize ram.

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-gprimeltecan.dts:
```diff
@@ -27,3 +27,42 @@
+		interrupt-parent = <&msmgpio>;
+		interrupts = <115 IRQ_TYPE_EDGE_RISING>;
+
+		vdd-supply = <&pm8916_l17>;
```
Looks like this should also be `&pm8916_l5` according to downstream:

https://github.com/msm8916-mainline/android_kernel_qcom_msm8916/blob/74c765e8c74df57881b6c94d4eda1b4d2a9ecda7/arch/arm/boot/dts/samsung/msm8916/msm8916-sec-fortuna-tmo-r02.dtsi#L199-L200

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-gprimeltecan.dts:
```diff
@@ -27,3 +27,42 @@
+		compatible = "bosch,bmc150_magn";
+		reg = <0x12>;
+
+		vdd-supply = <&pm8916_l17>;
```
same here

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna3g.dts:
```diff
@@ -42,6 +63,12 @@
 };
 
+&dsi0 {
+	panel@0 {
```
I'd prefer `&panel { ... }` on top level, rather than &dsi -> panel@0.

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8916-samsung-fortuna3g.dts:
```diff
@@ -15,6 +15,16 @@
+		gps_mem: gps@8c000000 {
+			reg = <0x0 0x8c000000 0x0 0x200000>;
+			no-map;
+		};
```
I'd drop this and put it in the common dtsi as written in an earlier comment.

stephan-gh reviewed (changes requested):
Looks good mostly, just some minor comments. :)

bolilingmeng89006 committed:
draft: arm64: dts: qcom: msm8916-samsung-gprimeltexx: Add initial device tree

Samsung Galaxy Grand Prime (SM-G530FZ) is a smartphone using the MSM8916
SoC released in 2014.

The Grand Prime variants are very similar, with some differences in
accelerometer and NFC. The common parts are shared in
msm8916-samsung-gprime-common.dtsi to reduce duplication.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

buddyjojo committed:
arm64: dts: qcom: msm8916-samsung-gprimeltecan: Add NFC

The Samsung Galaxy Grand Prime CAN has a Samsung S3FWRN5 NFC chip that
works quite well with the s3fwrn5 driver in the Linux NFC subsystem.

The clock setup for the NFC chip is a bit special (although this
seems to be a common approach used for Qualcomm devices with NFC):

The NFC chip has an output GPIO that is asserted whenever the clock
is needed to function properly. On the A3/A5 this is wired up to
PM8916 GPIO2, which is then configured with a special function
(NFC_CLK_REQ or BB_CLK2_REQ).

Enabling the rpmcc RPM_SMD_BB_CLK2_PIN clock will then instruct
PM8916 to automatically enable the clock whenever the NFC chip
requests it. The advantage is that the clock is only enabled when
needed and we don't need to manage it ourselves from the NFC driver.

Signed-off-by: Joe Mason <buddyjojo06@outlook.com>

Commit:
squash! MSM8916: drm/panel: Generate using linux-mdss-dsi-panel-driver-generator

X-Code-Generator: msm8916-mainline/linux-panel-drivers@90b1152
Signed-off-by: lmdpdg <lmdpdg@localhost>

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-fortunaltezt.dts:
```diff
@@ -0,0 +1,129 @@
+			no-map;
+		};
+
+		mpss_mem: mpss@86800000 {
+			reg = <0x0 0x86800000 0x0 0x5a00000>;
+			no-map;
+		};
```
Todo: Smaller value

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8916-samsung-gprimeltecan.dts:
```diff
@@ -0,0 +1,116 @@
+			no-map;
+		};
+
+		mpss_mem: mpss@86800000 {
+			reg = <0x0 0x86800000 0x0 0x5800000>;
+			no-map;
+		};
```
Todo: Smaller value

wonderfulShrineMaidenOfParadise commented on drivers/gpu/drm/panel/msm8916-generated/panel-samsung-hx8389c.c:
```diff
@@ -0,0 +1,388 @@
+	if (!ctx)
+		return -ENOMEM;
+
+	ctx->supplies[0].supply = "vreg";
```
```suggestion
	ctx->supplies[0].supply = "vddio";
```

Todo: Update l-p-d script

bolilingmeng89006 committed:
fixup! drm/panel: hx8389c: rename vreg to vddio

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
fixup! msm8916-samsung-fortunaltezt: modem and gps memory

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
Revert "arm64: dts: qcom: msm8916-samsung-fortunaltezt: Add display panel"

This reverts commit 3624f7ac5591ec47af3d3686d5d987239f77bcf6.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
Revert "squash! MSM8916: drm/panel: Generate using linux-mdss-dsi-panel-driver-generator"

This reverts commit c4c3b048c40a0be0cacae0ee18cf32f48c7b1d97.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
squash! MSM8916: drm/panel: Generate using linux-mdss-dsi-panel-driver-generator

X-Code-Generator: msm8916-mainline/linux-panel-drivers@0aec81e (dirty)
Signed-off-by: lmdpdg <lmdpdg@localhost>
Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
squash! MSM8916: drm/panel: Generate using linux-mdss-dsi-panel-driver-generator

X-Code-Generator: msm8916-mainline/linux-panel-drivers@cb74114 (dirty)
Signed-off-by: lmdpdg <lmdpdg@localhost>
Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

stephan-gh committed:
Revert "HACK: thermal/drivers/tsens: gprime"

This reverts commit 91ca60db0c6de7bc11322d312549eed9eb5cc2df.

stephan-gh committed:
gprime tsens test

SiddharthManthan commented:
I will leave generating the panel drivers to @wonderfulShrineMaidenOfParadise as I don't know what changes were made to the panel drivers.

stephan-gh committed:
iio: accel: st_accel: Add LSM303C

Signed-off-by: Stephan Gerhold <stephan@gerhold.net>

stephan-gh committed:
iio: magnetometer: st_magn: Add LSM303C

Signed-off-by: Stephan Gerhold <stephan@gerhold.net>

stephan-gh committed:
squash! arm64: dts: qcom: msm8916-samsung-fortuna3g: Add accelerometer

SiddharthManthan committed:
try: Magnetometer

stephan-gh committed:
fortunaltezt: try pwm backlight

stephan-gh commented:
See https://github.com/msm8916-mainline/linux/pull/228
